### PR TITLE
M-expression -> S-expression 

### PR DESCRIPTION
**Pull request by rocky**
@mmatera I reread the wiki information on M-expression again and what we have is an S-expression. 

When they refer to Mathematica what they are referring to is the what you type into it. Not how it is represented underneath which in Mathics (and probably Mathematica too) is an S-expression. The difference has to do with whether Head can be something other than an atom at the data structure level seen in the interpreter, not the form that you input things. 

If it is the case that Head can be another expression inside the Mathics interpreter, I'd be interested in learning of such examples. 

Also there are some changes based on recent confusions:
    
* M-expression back to S-expression.
* Document summary_text
* Correct and expand tutorial examples
* Clarify elements and element

### CODE REVIEW

**Review by TiagoCavalcante — APPROVED** [review]
LGTM

**Comment by mmatera** [comment]
@rocky thanks for looking at this. The first  example that comes to my mind are constructions with Derivative.
```
In[1]:= D[F[x,y],x,y]
Out[1]= Derivative[1, 1][F][x, y]
```
Here ``Derivative[1,1]`` is the head of ``Derivative[1,1][F]`` that is the head of ``Derivative[1, 1][F][x, y]``. There are other examples, but I think this is the simplest.

If you define then ``Derivative[1,1][F]:=g``, and reevaluate ``%1``, you get ``g[x,y]``
```
In[2]:= Derivative[1,1][F]:=g;
In[3]:= %1
Out[3] = g[x,y]
```

See:
https://reference.wolfram.com/language/ref/Derivative.html
and a couple of other examples:
https://reference.wolfram.com/language/ref/Map.html
https://reference.wolfram.com/language/ref/Key.html?q=Key

**Comment by rocky** [comment]
> @rocky thanks for looking at this. The first example that comes to my mind are constructions with Derivative.

Ok - great. What is a specific example? When run under `mathics --full-form` what is the output that appears?

(And then we also need to look at the Expression inside the interpreter)

File: docs/conf.py [hunk 1]
```diff
@@ -53,4 +53,4 @@
 # Add any paths that contain custom static files (such as style sheets) here,
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
+# html_static_path = ['_static']
```
Comment (mmatera) [inline docs/conf.py:56]: How is this related to the PR?

File: docs/extending/code-overview.rst [hunk 2]
```diff
@@ -3,15 +3,15 @@ Mathics Core Interpreter Overview
 =================================
 
 At its heart, the Mathics interpreter works over `M-expresssion
-<https://en.wikipedia.org/wiki/M-expression>`_: combinations of nested
+<https://en.wikipedia.org/wiki/S-expression>`_: combinations of nested
```
Comment (mmatera) [inline docs/extending/code-overview.rst]: I am still not convinced..

File: docs/extending/code-overview/glossary.rst [hunk 3]
```diff
@@ -175,6 +174,18 @@ Glossary
 
      For the general concept, see `Rewriting <https://en.wikipedia.org/wiki/Rewriting>`_.
 
+   S-expression
+
+     A structure which consists of a sequence atoms, and (nested)
+     expressions. However at each level there is a ``Head`` which
+     represents an operator or function.
```
Comment (mmatera) [inline docs/extending/code-overview/glossary.rst]: In ``S-Expressions`` the head is always an ``Atom``. But in WL, the head sometimes is another non-atomic expression, like in ``Derivative[1,2][f]``

File: docs/extending/code-overview.rst (hunk 2 above)
Comment (rocky) [reply docs/extending/code-overview.rst]: Ok - so find an example where this is not true.  If you want to demonstrate in the code where head is not assumed to be an Atom, I'd be grateful to learn. (I've been meaning to do this myself but haven't had a chance)

Also, on reflection even _if_ there turns out to be _rare_ situations where this isn't an S-expression, I am leaning towards using that word because it is much more commonly understood. So somewhere _if_ it happens to be a M-expression a footnote somewhere would point this out and just continue with using the more familiar and common idea.

File: docs/conf.py (hunk 1 above)
Comment (rocky) [reply docs/conf.py:56]: It isn't - sorry.  It was a warning noticed in build. I have now changed this in master and rebased.

File: docs/extending/code-overview/glossary.rst (hunk 3 above)
Comment (rocky) [reply docs/extending/code-overview/glossary.rst]: Ok. I just checked in the code, and yes, you are correct.  In rewrite_apply_eval_step `self._head` is:

```
<Expression: System`Derivative[1, 2]>
```

and that is why [line 923](https://github.com/Mathics3/mathics-core/blob/master/mathics/core/expression.py#L923) has the `.evaluate(evaluation)` call. 

Still I am not convinced using M-expression is the right way to go here rather than saying that this is really a _variant_ of an S-expression called an "M-expression" a few places in the docs.  

And the fact that we aren't turning this into an S-expression may also be a source of inefficiency. But as long as line 923 is there, we should make note that this is there because we have such kinds of situations and that this is where the distinction between S-expression and M-expression lies.

Comment (rocky) [reply docs/extending/code-overview/glossary.rst]: @mmatera In 6c3469e we have gone back to M-expression, while trying to make it clear that this is more or less the same thing as an S-expression so that people who are familiar with how things normally work, won't get confused by the slight variation here. 

Also in tha last commit, I start to define `Symbol`. I suspect that "binding" will ultimately get folded into "Definition", but that's for a different time and PR. 

Lastly, I find it amusing that you are incapable of giving the _simplest_ example  (e.g. `Derivative[1][f]`) that gets across the point. I would have thought in the education business this would be a useful thing to be able to be able do.

File: docs/extending/code-overview.rst [hunk 4]
```diff
@@ -2,16 +2,22 @@
...
+interpreters work, except the head element can be an expression rather
+than just a simple function name. See also `Everything is an
+Expression
 <https://reference.wolfram.com/language/tutorial/Expressions.html#4715>`_.
 
 When you enter a string to Mathics there is a 3-step process:
 
 * the input string is parsed to ``FullForm``
-* The ``FullForm`` S-expression is evaluated giving another S-expression
+* The ``FullForm`` M-expression is evaluated giving another M-expression
```
Comment (TiagoCavalcante) [inline docs/extending/code-overview.rst:20]: Weren't we going with S-Expression?

Comment (rocky) [reply docs/extending/code-overview.rst:20]: @mmatera correctly points out that the way things are implemented, what is seen in some cases are M-expressions. 

In particular when you enter `Derivative[1][f]`, `Derivitave[1]` acts like a single function name. 

So we are describing things as they currently appear in the code.  

A downside of of not using S-expressions which is the more commonly used kind of thing is that most of the literature and work done on lisp-like and functional-like evaluators assume this. So when there is some technology that comes up and we want to use, we'll have to look this over to see if it is amenable to M-expressions. 

I am pretty sure there is a way at the parser to eliminate this so that we do in fact S-expressions, but this isn't something we want to tackle right now.

Comment (mmatera) [reply docs/extending/code-overview.rst:20]: @rocky, as I told you, apart from the comments, I was OK with your changes. If you think that calling *S-Expression* to something that is actually *M-Expression*, for the sake of readability, is OK for me.  Later we can add a footnote clarifying this point.

File: docs/extending/code-overview/glossary.rst (hunk 3 above)
Comment (mmatera) [reply docs/extending/code-overview/glossary.rst]: @rocky, I had chosen carefully the "simplest" example that I imagined would avoid thinking that the expression would be equivalent to a single symbol "Derivative$1$f". In any case, it is clear that I failed :(.

Comment (rocky) [reply docs/extending/code-overview/glossary.rst]: No problem - please keep it in mind though. Practice makes perfect. And you may be able to use this idea or technique in the classroom. In my opinion, it was one of things that made Feynman a great lecturer.

And also, I am a very simple-minded person. I get confused with unnecessary distraction.
